// smart-webcomponents-angular/esm2020/threedchart/smart.threedchart.module.mjs
import { NgModule } from '@angular/core';
import { ThreeDChartComponent } from './smart.threedchart';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ThreeDChartModule {
}
ThreeDChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ThreeDChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ThreeDChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ThreeDChartModule, declarations: [ThreeDChartComponent], exports: [ThreeDChartComponent] });
ThreeDChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ThreeDChartModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ThreeDChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ThreeDChartComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ThreeDChartComponent]
                }]
        }] });
