// smart-webcomponents-angular/esm2020/colorpicker/smart.colorpicker.module.mjs
import { NgModule } from '@angular/core';
import { ColorPickerComponent } from './smart.colorpicker';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ColorPickerModule {
}
ColorPickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ColorPickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPickerModule, declarations: [ColorPickerComponent], exports: [ColorPickerComponent] });
ColorPickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPickerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ColorPickerComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ColorPickerComponent]
                }]
        }] });
