// smart-webcomponents-angular/esm2020/multiinput/smart.multiinput.module.mjs
import { NgModule } from '@angular/core';
import { MultiInputComponent } from './smart.multiinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class MultiInputModule {
}
MultiInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiInputModule, declarations: [MultiInputComponent], exports: [MultiInputComponent] });
MultiInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MultiInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MultiInputComponent]
                }]
        }] });
