// smart-webcomponents-angular/esm2020/accordion/smart.accordion.module.mjs
import { NgModule } from '@angular/core';
import { AccordionComponent } from './smart.accordion';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { AccordionItemComponent } from './smart.accordionitem';
import * as i0 from "@angular/core";
export class AccordionModule {
}
AccordionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: AccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AccordionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: AccordionModule, declarations: [AccordionComponent, AccordionItemComponent], exports: [AccordionComponent, AccordionItemComponent] });
AccordionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: AccordionModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: AccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AccordionComponent, AccordionItemComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [AccordionComponent, AccordionItemComponent]
                }]
        }] });
