// smart-webcomponents-angular/esm2020/tank/smart.tank.module.mjs
import { NgModule } from '@angular/core';
import { TankComponent } from './smart.tank';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class TankModule {
}
TankModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TankModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TankModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TankModule, declarations: [TankComponent], exports: [TankComponent] });
TankModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TankModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TankModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TankComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [TankComponent]
                }]
        }] });
