// smart-webcomponents-angular/esm2020/querybuilder/smart.querybuilder.module.mjs
import { NgModule } from '@angular/core';
import { QueryBuilderComponent } from './smart.querybuilder';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class QueryBuilderModule {
}
QueryBuilderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QueryBuilderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
QueryBuilderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QueryBuilderModule, declarations: [QueryBuilderComponent], exports: [QueryBuilderComponent] });
QueryBuilderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QueryBuilderModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QueryBuilderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [QueryBuilderComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [QueryBuilderComponent]
                }]
        }] });
