// smart-webcomponents-angular/esm2020/radiobutton/smart.radiobutton.module.mjs
import { NgModule } from '@angular/core';
import { RadioButtonComponent } from './smart.radiobutton';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class RadioButtonModule {
}
RadioButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RadioButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RadioButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RadioButtonModule, declarations: [RadioButtonComponent], exports: [RadioButtonComponent] });
RadioButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RadioButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RadioButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RadioButtonComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [RadioButtonComponent]
                }]
        }] });
