// smart-webcomponents-angular/esm2020/buttongroup/smart.buttongroup.module.mjs
import { NgModule } from '@angular/core';
import { ButtonGroupComponent } from './smart.buttongroup';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ButtonGroupModule {
}
ButtonGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ButtonGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonGroupModule, declarations: [ButtonGroupComponent], exports: [ButtonGroupComponent] });
ButtonGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonGroupModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ButtonGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ButtonGroupComponent]
                }]
        }] });
