// smart-webcomponents-angular/esm2020/switchbutton/smart.switchbutton.module.mjs
import { NgModule } from '@angular/core';
import { SwitchButtonComponent } from './smart.switchbutton';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class SwitchButtonModule {
}
SwitchButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SwitchButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SwitchButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SwitchButtonModule, declarations: [SwitchButtonComponent], exports: [SwitchButtonComponent] });
SwitchButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SwitchButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SwitchButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SwitchButtonComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [SwitchButtonComponent]
                }]
        }] });
