// smart-webcomponents-angular/esm2020/ganttchart/smart.ganttchart.module.mjs
import { NgModule } from '@angular/core';
import { GanttChartComponent } from './smart.ganttchart';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class GanttChartModule {
}
GanttChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GanttChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GanttChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GanttChartModule, declarations: [GanttChartComponent], exports: [GanttChartComponent] });
GanttChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GanttChartModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GanttChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [GanttChartComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [GanttChartComponent]
                }]
        }] });
