// smart-webcomponents-angular/esm2020/maskedtextbox/smart.maskedtextbox.module.mjs
import { NgModule } from '@angular/core';
import { MaskedTextBoxComponent } from './smart.maskedtextbox';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class MaskedTextBoxModule {
}
MaskedTextBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MaskedTextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MaskedTextBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MaskedTextBoxModule, declarations: [MaskedTextBoxComponent], exports: [MaskedTextBoxComponent] });
MaskedTextBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MaskedTextBoxModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MaskedTextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MaskedTextBoxComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MaskedTextBoxComponent]
                }]
        }] });
