// smart-webcomponents-angular/esm2020/led/smart.led.module.mjs
import { NgModule } from '@angular/core';
import { LedComponent } from './smart.led';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class LedModule {
}
LedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LedModule, declarations: [LedComponent], exports: [LedComponent] });
LedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LedModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LedModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LedComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [LedComponent]
                }]
        }] });
