// smart-webcomponents-angular/esm2020/multicomboinput/smart.multicomboinput.module.mjs
import { NgModule } from '@angular/core';
import { MultiComboInputComponent } from './smart.multicomboinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class MultiComboInputModule {
}
MultiComboInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiComboInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiComboInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiComboInputModule, declarations: [MultiComboInputComponent], exports: [MultiComboInputComponent] });
MultiComboInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiComboInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiComboInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MultiComboInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MultiComboInputComponent]
                }]
        }] });
