// smart-webcomponents-angular/esm2020/multilinetextbox/smart.multilinetextbox.module.mjs
import { NgModule } from '@angular/core';
import { MultilineTextBoxComponent } from './smart.multilinetextbox';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ListItemComponent } from './smart.listitem';
import { ListItemsGroupComponent } from './smart.listitemsgroup';
import * as i0 from "@angular/core";
export class MultilineTextBoxModule {
}
MultilineTextBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultilineTextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultilineTextBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultilineTextBoxModule, declarations: [MultilineTextBoxComponent, ListItemComponent, ListItemsGroupComponent], exports: [MultilineTextBoxComponent, ListItemComponent, ListItemsGroupComponent] });
MultilineTextBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultilineTextBoxModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultilineTextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MultilineTextBoxComponent, ListItemComponent, ListItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MultilineTextBoxComponent, ListItemComponent, ListItemsGroupComponent]
                }]
        }] });
