// smart-webcomponents-angular/esm2020/grouppanel/smart.grouppanel.module.mjs
import { NgModule } from '@angular/core';
import { GroupPanelComponent } from './smart.grouppanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class GroupPanelModule {
}
GroupPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GroupPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
GroupPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GroupPanelModule, declarations: [GroupPanelComponent], exports: [GroupPanelComponent] });
GroupPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GroupPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: GroupPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [GroupPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [GroupPanelComponent]
                }]
        }] });
