// smart-webcomponents-angular/esm2020/columnpanel/smart.columnpanel.module.mjs
import { NgModule } from '@angular/core';
import { ColumnPanelComponent } from './smart.columnpanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ColumnPanelModule {
}
ColumnPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColumnPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ColumnPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColumnPanelModule, declarations: [ColumnPanelComponent], exports: [ColumnPanelComponent] });
ColumnPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColumnPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColumnPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ColumnPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ColumnPanelComponent]
                }]
        }] });
