// smart-webcomponents-angular/esm2020/scrollbar/smart.scrollbar.module.mjs
import { NgModule } from '@angular/core';
import { ScrollBarComponent } from './smart.scrollbar';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ScrollBarModule {
}
ScrollBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ScrollBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ScrollBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ScrollBarModule, declarations: [ScrollBarComponent], exports: [ScrollBarComponent] });
ScrollBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ScrollBarModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ScrollBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ScrollBarComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ScrollBarComponent]
                }]
        }] });
