// smart-webcomponents-angular/esm2020/filterpanel/smart.filterpanel.module.mjs
import { NgModule } from '@angular/core';
import { FilterPanelComponent } from './smart.filterpanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class FilterPanelModule {
}
FilterPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilterPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterPanelModule, declarations: [FilterPanelComponent], exports: [FilterPanelComponent] });
FilterPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FilterPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [FilterPanelComponent]
                }]
        }] });
