// smart-webcomponents-angular/esm2020/button/smart.button.module.mjs
import { NgModule } from '@angular/core';
import { ButtonComponent } from './smart.button';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { RepeatButtonComponent } from './smart.repeatbutton';
import { ToggleButtonComponent } from './smart.togglebutton';
import { PowerButtonComponent } from './smart.powerbutton';
import * as i0 from "@angular/core";
export class ButtonModule {
}
ButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonModule, declarations: [ButtonComponent, RepeatButtonComponent, ToggleButtonComponent, PowerButtonComponent], exports: [ButtonComponent, RepeatButtonComponent, ToggleButtonComponent, PowerButtonComponent] });
ButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ButtonComponent, RepeatButtonComponent, ToggleButtonComponent, PowerButtonComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ButtonComponent, RepeatButtonComponent, ToggleButtonComponent, PowerButtonComponent]
                }]
        }] });
