// smart-webcomponents-angular/esm2020/progressbar/smart.progressbar.module.mjs
import { NgModule } from '@angular/core';
import { ProgressBarComponent } from './smart.progressbar';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CircularProgressBarComponent } from './smart.circularprogressbar';
import * as i0 from "@angular/core";
export class ProgressBarModule {
}
ProgressBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ProgressBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProgressBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ProgressBarModule, declarations: [ProgressBarComponent, CircularProgressBarComponent], exports: [ProgressBarComponent, CircularProgressBarComponent] });
ProgressBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ProgressBarModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ProgressBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ProgressBarComponent, CircularProgressBarComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ProgressBarComponent, CircularProgressBarComponent]
                }]
        }] });
