// smart-webcomponents-angular/esm2020/dropdownbutton/smart.dropdownbutton.module.mjs
import { NgModule } from '@angular/core';
import { DropDownButtonComponent } from './smart.dropdownbutton';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class DropDownButtonModule {
}
DropDownButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownButtonModule, declarations: [DropDownButtonComponent], exports: [DropDownButtonComponent] });
DropDownButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DropDownButtonComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DropDownButtonComponent]
                }]
        }] });
