// smart-webcomponents-angular/esm2020/tabs/smart.tabitem.mjs
import { Directive, Input } from '@angular/core';
import { BaseElement } from './smart.element';
import * as i0 from "@angular/core";
export { Smart } from './smart.element';
export class TabItemComponent extends BaseElement {
    constructor(ref) {
        super(ref);
        this.eventHandlers = [];
        this.nativeElement = ref.nativeElement;
    }
    /** @description Creates the component on demand.
     * @param properties An optional object of properties, which will be added to the template binded ones.
     */
    createComponent(properties = {}) {
        this.nativeElement = document.createElement('smart-tab-item');
        for (let propertyName in properties) {
            this.nativeElement[propertyName] = properties[propertyName];
        }
        return this.nativeElement;
    }
    /** @description Tab item close button state */
    get closeButtonHidden() {
        return this.nativeElement ? this.nativeElement.closeButtonHidden : undefined;
    }
    set closeButtonHidden(value) {
        this.nativeElement ? this.nativeElement.closeButtonHidden = value : undefined;
    }
    /** @description Disables the Tab item */
    get disabled() {
        return this.nativeElement ? this.nativeElement.disabled : undefined;
    }
    set disabled(value) {
        this.nativeElement ? this.nativeElement.disabled = value : undefined;
    }
    /** @description Tab item index */
    get index() {
        return this.nativeElement ? this.nativeElement.index : undefined;
    }
    set index(value) {
        this.nativeElement ? this.nativeElement.index = value : undefined;
    }
    /** @description Tab item selected state */
    get selected() {
        return this.nativeElement ? this.nativeElement.selected : undefined;
    }
    set selected(value) {
        this.nativeElement ? this.nativeElement.selected = value : undefined;
    }
    /** @description Tab item label */
    get label() {
        return this.nativeElement ? this.nativeElement.label : undefined;
    }
    set label(value) {
        this.nativeElement ? this.nativeElement.label = value : undefined;
    }
    /** @description Tab item content */
    get content() {
        return this.nativeElement ? this.nativeElement.content : undefined;
    }
    set content(value) {
        this.nativeElement ? this.nativeElement.content = value : undefined;
    }
    /** @description Tab item label size */
    get labelSize() {
        return this.nativeElement ? this.nativeElement.labelSize : undefined;
    }
    set labelSize(value) {
        this.nativeElement ? this.nativeElement.labelSize = value : undefined;
    }
    get isRendered() {
        return this.nativeElement ? this.nativeElement.isRendered : false;
    }
    ngOnInit() {
    }
    ngAfterViewInit() {
        const that = this;
        that.onCreate.emit(that.nativeElement);
        this.nativeElement.classList.add('smart-angular');
        this.nativeElement.whenRendered(() => { that.onReady.emit(that.nativeElement); });
    }
    ngOnDestroy() { }
    ngOnChanges(changes) {
        if (this.nativeElement && this.nativeElement.isRendered) {
            for (const propName in changes) {
                if (changes.hasOwnProperty(propName)) {
                    this.nativeElement[propName] = changes[propName].currentValue;
                }
            }
        }
    }
}
TabItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabItemComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TabItemComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.2.6", type: TabItemComponent, selector: "smart-tab-item, [smart-tab-item]", inputs: { closeButtonHidden: "closeButtonHidden", disabled: "disabled", index: "index", selected: "selected", label: "label", content: "content", labelSize: "labelSize" }, exportAs: ["smart-tab-item"], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabItemComponent, decorators: [{
            type: Directive,
            args: [{
                    exportAs: 'smart-tab-item', selector: 'smart-tab-item, [smart-tab-item]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { closeButtonHidden: [{
                type: Input
            }], disabled: [{
                type: Input
            }], index: [{
                type: Input
            }], selected: [{
                type: Input
            }], label: [{
                type: Input
            }], content: [{
                type: Input
            }], labelSize: [{
                type: Input
            }] } });
