// smart-webcomponents-angular/esm2020/tabs/smart.tabs.module.mjs
import { NgModule } from '@angular/core';
import { TabsComponent } from './smart.tabs';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { TabItemComponent } from './smart.tabitem';
import { TabItemsGroupComponent } from './smart.tabitemsgroup';
import * as i0 from "@angular/core";
export class TabsModule {
}
TabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabsModule, declarations: [TabsComponent, TabItemComponent, TabItemsGroupComponent], exports: [TabsComponent, TabItemComponent, TabItemsGroupComponent] });
TabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TabsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TabsComponent, TabItemComponent, TabItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [TabsComponent, TabItemComponent, TabItemsGroupComponent]
                }]
        }] });
