// smart-webcomponents-angular/esm2020/qrcode/smart.qrcode.module.mjs
import { NgModule } from '@angular/core';
import { QRcodeComponent } from './smart.qrcode';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class QRcodeModule {
}
QRcodeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QRcodeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
QRcodeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QRcodeModule, declarations: [QRcodeComponent], exports: [QRcodeComponent] });
QRcodeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QRcodeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: QRcodeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [QRcodeComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [QRcodeComponent]
                }]
        }] });
