// smart-webcomponents-angular/esm2020/textbox/smart.textbox.module.mjs
import { NgModule } from '@angular/core';
import { TextBoxComponent } from './smart.textbox';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ListItemComponent } from './smart.listitem';
import { ListItemsGroupComponent } from './smart.listitemsgroup';
import * as i0 from "@angular/core";
export class TextBoxModule {
}
TextBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TextBoxModule, declarations: [TextBoxComponent, ListItemComponent, ListItemsGroupComponent], exports: [TextBoxComponent, ListItemComponent, ListItemsGroupComponent] });
TextBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TextBoxModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TextBoxComponent, ListItemComponent, ListItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [TextBoxComponent, ListItemComponent, ListItemsGroupComponent]
                }]
        }] });
