// smart-webcomponents-angular/esm2020/form/smart.form.module.mjs
import { NgModule } from '@angular/core';
import { FormComponent } from './smart.form';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { FormControlComponent } from './smart.formcontrol';
import { FormGroupComponent } from './smart.formgroup';
import * as i0 from "@angular/core";
export class FormModule {
}
FormModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FormModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FormModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FormModule, declarations: [FormComponent, FormControlComponent, FormGroupComponent], exports: [FormComponent, FormControlComponent, FormGroupComponent] });
FormModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FormModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FormModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FormComponent, FormControlComponent, FormGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [FormComponent, FormControlComponent, FormGroupComponent]
                }]
        }] });
