// smart-webcomponents-angular/esm2020/numberinput/smart.numberinput.module.mjs
import { NgModule } from '@angular/core';
import { NumberInputComponent } from './smart.numberinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class NumberInputModule {
}
NumberInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NumberInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NumberInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NumberInputModule, declarations: [NumberInputComponent], exports: [NumberInputComponent] });
NumberInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NumberInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NumberInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NumberInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [NumberInputComponent]
                }]
        }] });
