// smart-webcomponents-angular/esm2020/countryinput/smart.countryinput.module.mjs
import { NgModule } from '@angular/core';
import { CountryInputComponent } from './smart.countryinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class CountryInputModule {
}
CountryInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CountryInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CountryInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CountryInputModule, declarations: [CountryInputComponent], exports: [CountryInputComponent] });
CountryInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CountryInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CountryInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CountryInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [CountryInputComponent]
                }]
        }] });
