// smart-webcomponents-angular/esm2020/barcode/smart.barcode.module.mjs
import { NgModule } from '@angular/core';
import { BarcodeComponent } from './smart.barcode';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class BarcodeModule {
}
BarcodeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BarcodeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BarcodeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BarcodeModule, declarations: [BarcodeComponent], exports: [BarcodeComponent] });
BarcodeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BarcodeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BarcodeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BarcodeComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [BarcodeComponent]
                }]
        }] });
