// smart-webcomponents-angular/esm2020/cardview/smart.cardview.module.mjs
import { NgModule } from '@angular/core';
import { CardViewComponent } from './smart.cardview';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class CardViewModule {
}
CardViewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CardViewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CardViewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CardViewModule, declarations: [CardViewComponent], exports: [CardViewComponent] });
CardViewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CardViewModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CardViewModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CardViewComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [CardViewComponent]
                }]
        }] });
