// smart-webcomponents-angular/esm2020/dateinput/smart.dateinput.module.mjs
import { NgModule } from '@angular/core';
import { DateInputComponent } from './smart.dateinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class DateInputModule {
}
DateInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateInputModule, declarations: [DateInputComponent], exports: [DateInputComponent] });
DateInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DateInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DateInputComponent]
                }]
        }] });
