// smart-webcomponents-angular/esm2020/editor/smart.editor.module.mjs
import { NgModule } from '@angular/core';
import { EditorComponent } from './smart.editor';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class EditorModule {
}
EditorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: EditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
EditorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: EditorModule, declarations: [EditorComponent], exports: [EditorComponent] });
EditorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: EditorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: EditorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [EditorComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [EditorComponent]
                }]
        }] });
