// smart-webcomponents-angular/esm2020/datetimepicker/smart.datetimepicker.module.mjs
import { NgModule } from '@angular/core';
import { DateTimePickerComponent } from './smart.datetimepicker';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class DateTimePickerModule {
}
DateTimePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateTimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateTimePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateTimePickerModule, declarations: [DateTimePickerComponent], exports: [DateTimePickerComponent] });
DateTimePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateTimePickerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateTimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DateTimePickerComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DateTimePickerComponent]
                }]
        }] });
