// smart-webcomponents-angular/esm2020/layout/smart.layout.module.mjs
import { NgModule } from '@angular/core';
import { LayoutComponent } from './smart.layout';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { LayoutItemComponent } from './smart.layoutitem';
import { LayoutGroupComponent } from './smart.layoutgroup';
import { TabLayoutItemComponent } from './smart.tablayoutitem';
import { TabLayoutGroupComponent } from './smart.tablayoutgroup';
import * as i0 from "@angular/core";
export class LayoutModule {
}
LayoutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LayoutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LayoutModule, declarations: [LayoutComponent, LayoutItemComponent, LayoutGroupComponent, TabLayoutItemComponent, TabLayoutGroupComponent], exports: [LayoutComponent, LayoutItemComponent, LayoutGroupComponent, TabLayoutItemComponent, TabLayoutGroupComponent] });
LayoutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LayoutModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: LayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LayoutComponent, LayoutItemComponent, LayoutGroupComponent, TabLayoutItemComponent, TabLayoutGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [LayoutComponent, LayoutItemComponent, LayoutGroupComponent, TabLayoutItemComponent, TabLayoutGroupComponent]
                }]
        }] });
