// smart-webcomponents-angular/esm2020/checkinput/smart.checkinput.module.mjs
import { NgModule } from '@angular/core';
import { CheckInputComponent } from './smart.checkinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class CheckInputModule {
}
CheckInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CheckInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CheckInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CheckInputModule, declarations: [CheckInputComponent], exports: [CheckInputComponent] });
CheckInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CheckInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CheckInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CheckInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [CheckInputComponent]
                }]
        }] });
