// smart-webcomponents-angular/esm2020/menu/smart.menu.module.mjs
import { NgModule } from '@angular/core';
import { MenuComponent } from './smart.menu';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { MenuItemComponent } from './smart.menuitem';
import { MenuItemsGroupComponent } from './smart.menuitemsgroup';
import * as i0 from "@angular/core";
export class MenuModule {
}
MenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MenuModule, declarations: [MenuComponent, MenuItemComponent, MenuItemsGroupComponent], exports: [MenuComponent, MenuItemComponent, MenuItemsGroupComponent] });
MenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MenuModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MenuModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MenuComponent, MenuItemComponent, MenuItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MenuComponent, MenuItemComponent, MenuItemsGroupComponent]
                }]
        }] });
