// smart-webcomponents-angular/esm2020/rating/smart.rating.module.mjs
import { NgModule } from '@angular/core';
import { RatingComponent } from './smart.rating';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class RatingModule {
}
RatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RatingModule, declarations: [RatingComponent], exports: [RatingComponent] });
RatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RatingModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: RatingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RatingComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [RatingComponent]
                }]
        }] });
