// smart-webcomponents-angular/esm2020/splitter/smart.splitter.module.mjs
import { NgModule } from '@angular/core';
import { SplitterComponent } from './smart.splitter';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { SplitterItemComponent } from './smart.splitteritem';
import { SplitterBarComponent } from './smart.splitterbar';
import * as i0 from "@angular/core";
export class SplitterModule {
}
SplitterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SplitterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SplitterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SplitterModule, declarations: [SplitterComponent, SplitterItemComponent, SplitterBarComponent], exports: [SplitterComponent, SplitterItemComponent, SplitterBarComponent] });
SplitterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SplitterModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SplitterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SplitterComponent, SplitterItemComponent, SplitterBarComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [SplitterComponent, SplitterItemComponent, SplitterBarComponent]
                }]
        }] });
