// smart-webcomponents-angular/esm2020/daterangeinput/smart.daterangeinput.module.mjs
import { NgModule } from '@angular/core';
import { DateRangeInputComponent } from './smart.daterangeinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class DateRangeInputModule {
}
DateRangeInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateRangeInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateRangeInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateRangeInputModule, declarations: [DateRangeInputComponent], exports: [DateRangeInputComponent] });
DateRangeInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateRangeInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DateRangeInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DateRangeInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DateRangeInputComponent]
                }]
        }] });
