// smart-webcomponents-angular/esm2020/colorinput/smart.colorinput.module.mjs
import { NgModule } from '@angular/core';
import { ColorInputComponent } from './smart.colorinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ColorInputModule {
}
ColorInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ColorInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorInputModule, declarations: [ColorInputComponent], exports: [ColorInputComponent] });
ColorInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ColorInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ColorInputComponent]
                }]
        }] });
