// smart-webcomponents-angular/esm2020/colorpanel/smart.colorpanel.module.mjs
import { NgModule } from '@angular/core';
import { ColorPanelComponent } from './smart.colorpanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ColorPanelModule {
}
ColorPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ColorPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPanelModule, declarations: [ColorPanelComponent], exports: [ColorPanelComponent] });
ColorPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ColorPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ColorPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ColorPanelComponent]
                }]
        }] });
