// smart-webcomponents-angular/esm2020/dropdownlist/smart.dropdownlist.module.mjs
import { NgModule } from '@angular/core';
import { DropDownListComponent } from './smart.dropdownlist';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ListItemComponent } from './smart.listitem';
import { ListItemsGroupComponent } from './smart.listitemsgroup';
import * as i0 from "@angular/core";
export class DropDownListModule {
}
DropDownListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownListModule, declarations: [DropDownListComponent, ListItemComponent, ListItemsGroupComponent], exports: [DropDownListComponent, ListItemComponent, ListItemsGroupComponent] });
DropDownListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownListModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DropDownListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DropDownListComponent, ListItemComponent, ListItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DropDownListComponent, ListItemComponent, ListItemsGroupComponent]
                }]
        }] });
