// smart-webcomponents-angular/esm2020/customizationdialog/smart.customizationdialog.module.mjs
import { NgModule } from '@angular/core';
import { CustomizationDialogComponent } from './smart.customizationdialog';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class CustomizationDialogModule {
}
CustomizationDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CustomizationDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CustomizationDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CustomizationDialogModule, declarations: [CustomizationDialogComponent], exports: [CustomizationDialogComponent] });
CustomizationDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CustomizationDialogModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: CustomizationDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CustomizationDialogComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [CustomizationDialogComponent]
                }]
        }] });
