// smart-webcomponents-angular/esm2020/phoneinput/smart.phoneinput.module.mjs
import { NgModule } from '@angular/core';
import { PhoneInputComponent } from './smart.phoneinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class PhoneInputModule {
}
PhoneInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PhoneInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PhoneInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PhoneInputModule, declarations: [PhoneInputComponent], exports: [PhoneInputComponent] });
PhoneInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PhoneInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PhoneInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PhoneInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [PhoneInputComponent]
                }]
        }] });
