// smart-webcomponents-angular/esm2020/pivottable/smart.pivottable.module.mjs
import { NgModule } from '@angular/core';
import { PivotTableComponent } from './smart.pivottable';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class PivotTableModule {
}
PivotTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PivotTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PivotTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PivotTableModule, declarations: [PivotTableComponent], exports: [PivotTableComponent] });
PivotTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PivotTableModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PivotTableModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PivotTableComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [PivotTableComponent]
                }]
        }] });
