// smart-webcomponents-angular/esm2020/kanban/smart.kanban.module.mjs
import { NgModule } from '@angular/core';
import { KanbanComponent } from './smart.kanban';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class KanbanModule {
}
KanbanModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: KanbanModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
KanbanModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: KanbanModule, declarations: [KanbanComponent], exports: [KanbanComponent] });
KanbanModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: KanbanModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: KanbanModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [KanbanComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [KanbanComponent]
                }]
        }] });
