// smart-webcomponents-angular/esm2020/listmenu/smart.listmenu.module.mjs
import { NgModule } from '@angular/core';
import { ListMenuComponent } from './smart.listmenu';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { MenuItemComponent } from './smart.menuitem';
import { MenuItemsGroupComponent } from './smart.menuitemsgroup';
import * as i0 from "@angular/core";
export class ListMenuModule {
}
ListMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ListMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ListMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ListMenuModule, declarations: [ListMenuComponent, MenuItemComponent, MenuItemsGroupComponent], exports: [ListMenuComponent, MenuItemComponent, MenuItemsGroupComponent] });
ListMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ListMenuModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ListMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ListMenuComponent, MenuItemComponent, MenuItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ListMenuComponent, MenuItemComponent, MenuItemsGroupComponent]
                }]
        }] });
