// smart-webcomponents-angular/esm2020/filterbuilder/smart.filterbuilder.module.mjs
import { NgModule } from '@angular/core';
import { FilterBuilderComponent } from './smart.filterbuilder';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class FilterBuilderModule {
}
FilterBuilderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterBuilderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilterBuilderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterBuilderModule, declarations: [FilterBuilderComponent], exports: [FilterBuilderComponent] });
FilterBuilderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterBuilderModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: FilterBuilderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FilterBuilderComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [FilterBuilderComponent]
                }]
        }] });
