// smart-webcomponents-angular/esm2020/validator/smart.validator.mjs
import { Directive, Input } from '@angular/core';
import { BaseElement, Smart } from './smart.element';
import * as i0 from "@angular/core";
export { Smart } from './smart.element';
export class ValidatorComponent extends BaseElement {
    constructor(ref) {
        super(ref);
        this.eventHandlers = [];
        this.nativeElement = ref.nativeElement;
    }
    /** @description Creates the component on demand.
     * @param properties An optional object of properties, which will be added to the template binded ones.
     */
    createComponent(properties = {}) {
        this.nativeElement = document.createElement('jqxValidator');
        for (let propertyName in properties) {
            this.nativeElement[propertyName] = properties[propertyName];
        }
        return this.nativeElement;
    }
    /** @description  */
    get rules() {
        return this.nativeElement ? this.nativeElement.rules : undefined;
    }
    set rules(value) {
        this.nativeElement ? this.nativeElement.rules = value : undefined;
    }
    /** @description A valid CSS selector of an element on the page to be used as a container for validation error messages. */
    get validationSummarySelector() {
        return this.nativeElement ? this.nativeElement.validationSummarySelector : undefined;
    }
    set validationSummarySelector(value) {
        this.nativeElement ? this.nativeElement.validationSummarySelector = value : undefined;
    }
    /** @description Clears the error messages.
    */
    reset() {
        if (this.nativeElement.isRendered) {
            this.nativeElement.reset();
        }
        else {
            this.nativeElement.whenRendered(() => {
                this.nativeElement.reset();
            });
        }
    }
    /** @description Opens the dropDown.
    * @param {Function} result?. A callback function to call when validating inputs.
    */
    validate(result) {
        if (this.nativeElement.isRendered) {
            this.nativeElement.validate(result);
        }
        else {
            this.nativeElement.whenRendered(() => {
                this.nativeElement.validate(result);
            });
        }
    }
    get isRendered() {
        return this.nativeElement ? this.nativeElement.isRendered : false;
    }
    ngOnInit() {
    }
    ngAfterViewInit() {
        const that = this;
        that.onCreate.emit(that.nativeElement);
        Smart.Render();
        this.nativeElement.classList.add('smart-angular');
        this.nativeElement.whenRendered(() => { that.onReady.emit(that.nativeElement); });
        this.listen();
    }
    ngOnDestroy() {
        this.unlisten();
    }
    ngOnChanges(changes) {
        if (this.nativeElement && this.nativeElement.isRendered) {
            for (const propName in changes) {
                if (changes.hasOwnProperty(propName)) {
                    this.nativeElement[propName] = changes[propName].currentValue;
                }
            }
        }
    }
    /** @description Add event listeners. */
    listen() {
        const that = this;
    }
    /** @description Remove event listeners. */
    unlisten() {
        const that = this;
    }
}
ValidatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorComponent, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
ValidatorComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.2.6", type: ValidatorComponent, selector: "jqxValidator, [jqxValidator]", inputs: { rules: "rules", validationSummarySelector: "validationSummarySelector" }, exportAs: ["jqxValidator"], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorComponent, decorators: [{
            type: Directive,
            args: [{
                    exportAs: 'jqxValidator', selector: 'jqxValidator, [jqxValidator]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { rules: [{
                type: Input
            }], validationSummarySelector: [{
                type: Input
            }] } });
