// smart-webcomponents-angular/esm2020/validator/smart.validator.module.mjs
import { NgModule } from '@angular/core';
import { ValidatorComponent } from './smart.validator';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ValidatorModule {
}
ValidatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ValidatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorModule, declarations: [ValidatorComponent], exports: [ValidatorComponent] });
ValidatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ValidatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ValidatorComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ValidatorComponent]
                }]
        }] });
