// smart-webcomponents-angular/esm2020/dockinglayout/smart.dockinglayout.module.mjs
import { NgModule } from '@angular/core';
import { DockingLayoutComponent } from './smart.dockinglayout';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class DockingLayoutModule {
}
DockingLayoutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DockingLayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DockingLayoutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DockingLayoutModule, declarations: [DockingLayoutComponent], exports: [DockingLayoutComponent] });
DockingLayoutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DockingLayoutModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DockingLayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DockingLayoutComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [DockingLayoutComponent]
                }]
        }] });
