// smart-webcomponents-angular/esm2020/path/smart.path.module.mjs
import { NgModule } from '@angular/core';
import { PathComponent } from './smart.path';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class PathModule {
}
PathModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PathModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PathModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PathModule, declarations: [PathComponent], exports: [PathComponent] });
PathModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PathModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PathModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PathComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [PathComponent]
                }]
        }] });
