// smart-webcomponents-angular/esm2020/passwordtextbox/smart.passwordtextbox.module.mjs
import { NgModule } from '@angular/core';
import { PasswordTextBoxComponent } from './smart.passwordtextbox';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class PasswordTextBoxModule {
}
PasswordTextBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordTextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PasswordTextBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordTextBoxModule, declarations: [PasswordTextBoxComponent], exports: [PasswordTextBoxComponent] });
PasswordTextBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordTextBoxModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordTextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PasswordTextBoxComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [PasswordTextBoxComponent]
                }]
        }] });
