// smart-webcomponents-angular/esm2020/sortpanel/smart.sortpanel.module.mjs
import { NgModule } from '@angular/core';
import { SortPanelComponent } from './smart.sortpanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class SortPanelModule {
}
SortPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SortPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SortPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SortPanelModule, declarations: [SortPanelComponent], exports: [SortPanelComponent] });
SortPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SortPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SortPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SortPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [SortPanelComponent]
                }]
        }] });
