// smart-webcomponents-angular/fesm2015/smart-webcomponents-angular.mjs
import * as i0 from '@angular/core';
import { Component, NgModule } from '@angular/core';

class DeleteComponent {
}
DeleteComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
DeleteComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.6", type: DeleteComponent, selector: "app-delete", ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'app-delete',
                    template: '<ng-content></ng-content>'
                }]
        }] });

class DeleteModule {
}
DeleteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeleteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteModule, declarations: [DeleteComponent], exports: [DeleteComponent] });
DeleteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: DeleteModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DeleteComponent],
                    exports: [DeleteComponent]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { DeleteComponent, DeleteModule };
