// smart-webcomponents-angular/esm2020/multisplitbutton/smart.multisplitbutton.module.mjs
import { NgModule } from '@angular/core';
import { MultiSplitButtonComponent } from './smart.multisplitbutton';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ListItemComponent } from './smart.listitem';
import { ListItemsGroupComponent } from './smart.listitemsgroup';
import * as i0 from "@angular/core";
export class MultiSplitButtonModule {
}
MultiSplitButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiSplitButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiSplitButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiSplitButtonModule, declarations: [MultiSplitButtonComponent, ListItemComponent, ListItemsGroupComponent], exports: [MultiSplitButtonComponent, ListItemComponent, ListItemsGroupComponent] });
MultiSplitButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiSplitButtonModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiSplitButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MultiSplitButtonComponent, ListItemComponent, ListItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MultiSplitButtonComponent, ListItemComponent, ListItemsGroupComponent]
                }]
        }] });
