// smart-webcomponents-angular/esm2020/passwordinput/smart.passwordinput.module.mjs
import { NgModule } from '@angular/core';
import { PasswordInputComponent } from './smart.passwordinput';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class PasswordInputModule {
}
PasswordInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PasswordInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordInputModule, declarations: [PasswordInputComponent], exports: [PasswordInputComponent] });
PasswordInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordInputModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: PasswordInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PasswordInputComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [PasswordInputComponent]
                }]
        }] });
