// smart-webcomponents-angular/esm2020/array/smart.array.module.mjs
import { NgModule } from '@angular/core';
import { ArrayComponent } from './smart.array';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class ArrayModule {
}
ArrayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ArrayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ArrayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ArrayModule, declarations: [ArrayComponent], exports: [ArrayComponent] });
ArrayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ArrayModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: ArrayModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ArrayComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [ArrayComponent]
                }]
        }] });
