// smart-webcomponents-angular/esm2020/tree/smart.tree.module.mjs
import { NgModule } from '@angular/core';
import { TreeComponent } from './smart.tree';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { TreeItemComponent } from './smart.treeitem';
import { TreeItemsGroupComponent } from './smart.treeitemsgroup';
import * as i0 from "@angular/core";
export class TreeModule {
}
TreeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TreeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TreeModule, declarations: [TreeComponent, TreeItemComponent, TreeItemsGroupComponent], exports: [TreeComponent, TreeItemComponent, TreeItemsGroupComponent] });
TreeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TreeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: TreeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TreeComponent, TreeItemComponent, TreeItemsGroupComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [TreeComponent, TreeItemComponent, TreeItemsGroupComponent]
                }]
        }] });
