// smart-webcomponents-angular/esm2020/multicolumnfilterpanel/smart.multicolumnfilterpanel.module.mjs
import { NgModule } from '@angular/core';
import { MultiColumnFilterPanelComponent } from './smart.multicolumnfilterpanel';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class MultiColumnFilterPanelModule {
}
MultiColumnFilterPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiColumnFilterPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiColumnFilterPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiColumnFilterPanelModule, declarations: [MultiColumnFilterPanelComponent], exports: [MultiColumnFilterPanelComponent] });
MultiColumnFilterPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiColumnFilterPanelModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: MultiColumnFilterPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MultiColumnFilterPanelComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [MultiColumnFilterPanelComponent]
                }]
        }] });
